daemon/config: support alternate (common) unicode encodings using a BOM

neersighted opened the pull request:
- Stacked on top of https://github.com/moby/moby/pull/44773
- Follow-up to https://github.com/moby/moby/pull/44770


**- What I did**
Instead of merely stripping a UTF-8 BOM, interpret UTF-8 and UTF-16 BOMs in order to support the common encodings on Windows.

This is a pragmatic but impure choice, in order to better support the default tools available on Windows Server, and reduce user confusion due to otherwise inscrutable-to-the-uninitiated errors like the following:

> invalid character 'þ' looking for beginning of value
> invalid character 'ÿ' looking for beginning of value

While meaningful to those who are familiar with and are equipped to diagnose encoding issues, these characters will be hidden when the file is edited with a BOM-aware text editor, and further confuse the user.

**- How I did it**

* https://pkg.go.dev/golang.org/x/text/transform
* https://pkg.go.dev/golang.org/x/text/encoding
* https://pkg.go.dev/golang.org/x/text/encoding/unicode

**- How to verify it**
Unit tests

**- Description for the changelog**
* Moby now supports UTF-16 encoded JSON configuration when a byte order mark is present
* Moby will report invalid UTF-8 in the JSON configuration early and comprehensibly

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/1049222/211658378-67915e03-faa7-47b2-8e07-c5ca73511f11.png)

neersighted committed:
daemon/config: clean up tests to use common helper

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

neersighted committed:
daemon/config: support alternate (common) unicode encodings using a BOM

This is a pragmatic but impure choice, in order to better support the
default tools available on Windows Server, and reduce user confusion due
to otherwise inscrutable-to-the-uninitiated errors like the following:

> invalid character 'þ' looking for beginning of value
> invalid character 'ÿ' looking for beginning of value

While meaningful to those who are familiar with and are equipped to
diagnose encoding issues, these characters will be hidden when the file
is edited with a BOM-aware text editor, and further confuse the user.

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

neersighted commented on daemon/config/config.go:
```diff
@@ -410,10 +412,36 @@ func getConflictFreeConfiguration(configFile string, flags *pflag.FlagSet) (*Con
+	// [Windows PowerShell]: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-5.1
+	decoder := unicode.BOMOverride(unicode.UTF8.NewDecoder())
+	b, _, err = transform.Bytes(decoder, b)
+	if err != nil {
```
This change also has the knock-on benefit that invalid UTF-8 should fail earlier, and in something that is obviously a encoding-related step, instead of causing a tokenization/other parse issue in the JSON parser.

corhere replied:
Will it? [The godoc for `encoding.Decoder`](https://pkg.go.dev/golang.org/x/text/encoding#Decoder) seems to disagree.

> Transforming source bytes that are not of that encoding will not result in an error per se. Each byte that cannot be transcoded will be represented in the output by the UTF-8 encoding of '\uFFFD', the replacement rune.

Sounds like good fodder for a unit test, regardless.

corhere replied:
[The implementation of `unicode.BOMOverride` will use `transform.Nop`](https://cs.opensource.google/go/x/text/+/refs/tags/v0.6.0:encoding/unicode/override.go;l=70-75) if it detects a UTF-8 BOM, so maybe this'd work?
```go
decoder := transformer.Chain(unicode.BOMOverride(transformer.Nop), encoding.UTF8Validator)
```

thaJeztah commented on daemon/config/config_test.go:
```diff
@@ -13,41 +13,68 @@ import (
+			encoding: unicode.UTF8,
+		},
+		{
+			name:     "UTF-8 (with BOM)",
+			encoding: unicode.UTF8BOM,
+		},
+		{
+			name:     "UTF-16 (BE with BOM)",
+			encoding: unicode.UTF16(unicode.BigEndian, unicode.UseBOM),
+		},
+		{
+			name:     "UTF-16 (LE with BOM)",
+			encoding: unicode.UTF16(unicode.LittleEndian, unicode.UseBOM),
+		},
```
I wonder if we're taking this a bit too far; would it be reasonable to tell that the file is expected to be in UTF-8? We have many JSON files, and don't handle them with the same assumptions.

(Als not a huge fan of introducing 3K+ lines of dependency, more so because there's been a lot of complexity in that dependency, and it had some vulnerabilities in the past).

corhere replied:
The daemon _writes_ many JSON files, and can therefore reasonably assume they will remain in the same encoding as written. The daemon config is a special case of a file which the user creates and the daemon only reads.

thaJeztah replied:
Right, we should document what format we expect; I think it would be reasonable to document "must be UTF-8, and a BOM is allowed".

ISTR we strip the BOM for Dockerfiles as well, but don't do other conversions beyond that.

tianon replied:
I tend to agree with Seb here -- `TrimBytes` is really unobtrusive (fully stdlib!) and covers the most common cases of failure here, but it feels a little excessive to bring in such a large extra dependency just to enable users to not have to care what their editor does to the config file (and even `notepad` has gotten better about not doing this recently).

neersighted replied:
The main reason for this/the issue I see is Windows Server; it simply is **hard** to write UTF-8 with no BOM on Windows Server 2019. Specifically, it:

* does not allow writing UTF-8 with no BOM in Notepad.exe; one must bail out to 'ANSI' and write UTF-8 escape sequences
* PowerShell does not support UTF-8 with no BOM except in [one Cmdlet](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-7.3#character-encoding-in-windows-powershell), `New-Item -Type File -Value`
* defaults to UTF-16LE or the system encoding in many places; when the system encoding is UTF-8, a BOM is often written

I think this is different than the JSON API or other interfaces to the Daemon as many of these tools are **expected ways to interact with the daemon**. Writing a Dockerfile is different as typically it is done with developer tools e.g. feature-rich text editors; the common tool to interact with `daemon.json` is PowerShell or Notepad, as the only text editor available out of the box on Windows.

We can certainly detect UTF-16 and yell at the user in a useful manner, which helps address the inscrutable errors I'm looking to improve here; however, that's still somewhat user-hostile as PowerShell's default encoding for `>`, `>>`, `Out-File`, and friends will be UTF-16.

I don't think extra dependencies is a huge concern here; we bring in `golang.org/x/text` and `golang.org/x/transform` already, and these are well-known, well-used, and well-audited libraries.

neersighted replied:
I've posted a stdlib-only alternative at https://github.com/moby/moby/pull/44796; it's not quite as user friendly, and I still prefer this approach.

To add a little more signal: this change is driven by feedback/conversations with the Mirantis support org, as well as an examination of customer questions and support cases, and having to write a 500-word article on what tools can safely manipulate daemon.json on a barebones Windows Server install.

I think one way or another we should improve here, and the other PR gets us most of the way there; however, I still would like to reduce confusion and frustration by making the operating system standard tools "just work," more or less.

neersighted committed:
daemon/config: inform the user when the input JSON contains invalid UTF-8

This helps ensure that users are not surprised by unexpected tokens in
the JSON parser, or fallout later in the daemon.

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

corhere commented on daemon/config/config.go:
```diff
@@ -410,10 +414,36 @@ func getConflictFreeConfiguration(configFile string, flags *pflag.FlagSet) (*Con
+	// [RFC 8259 Section 8.1]: https://www.rfc-editor.org/rfc/rfc8259#section-8.1
+	// [RFC 7159 Section 8.1]: https://www.rfc-editor.org/rfc/rfc7159#section-8.1
+	// [Windows PowerShell]: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-5.1
+	decoder := unicode.BOMOverride(unicode.UTF8.NewDecoder())
```
```suggestion
	decoder := unicode.BOMOverride(transform.Nop)
```
`UTF8.NewDecoder()` may replace invalid characters with replacement codepoints, so the chained validator would always see valid UTF-8 unless the file contained a UTF-8 BOM.

neersighted replied:
Ah, good catch, I need to widen my test-case a bit it seems.

corhere commented on daemon/config/config.go:
```diff
@@ -410,10 +414,35 @@ func getConflictFreeConfiguration(configFile string, flags *pflag.FlagSet) (*Con
+	// [Windows PowerShell]: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-5.1
+	b, n, err := transform.Bytes(transform.Chain(unicode.BOMOverride(transform.Nop), encoding.UTF8Validator), b)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to decode configuration JSON at character %d", n)
```
```suggestion
		return nil, errors.Wrapf(err, "failed to decode configuration JSON at offset %d", n)
```
"Character" is a rather ill-defined concept in the world of invalid text.

neersighted replied:
Fair, "byte" or "offset" are much more meaningful here (yay, variable width encodings!).

tianon reviewed (approved):
After a very productive (and generally shared session of sadness) maintainer call, I'm convinced this is probably the best way forward (and no build tag gate in the name of simpler code flow / future maintenance).

As discussed in the call, it'd be neat to have some explicit note about what conditions we're looking for to be able to remove this code (notably, Windows no longer commonly creating BOMs with included tools like Powershell by default), but I think the existing long comment here is pretty OK as-is too. :+1:

thaJeztah commented:
Post-merge LGTM

To capture some additional bits that were discussed;


- The `docker` CLI also uses JSON for configuration (usually `~/.docker/config.json`). While this file may be "less common" to be generated/manipulated with automation (PowerShell CmdLets), it _may_ be possible that users run into this, and if we receive a strong indication that this is causing issues for Windows users, we can consider giving that file the same treatment.
- As outlined in the comment added in this PR, we do _NOT_ consider extending beyond that scope (unless there's a very strong reason); JSON _should_ be UTF-8 without BOM, and we prefer to consider that our contract as well.

Adding to all of the above;

- The current choice of using JSON for these configuration files is not _ideal_. While JSON has advantages for some things (well understood format, can be serialized to a 1-line string), it has many _disadvantages_ for "humans" (difficult to manipulate, no option for comments, etc.)
- Combined with the fact that we only use a _single_ JSON file, this makes configuration difficult.

We need to improve this (both on the CLI and daemon side); linking some issues related to that (but there's more);

- https://github.com/moby/moby/issues/20151
- https://github.com/moby/moby/issues/44640
